File: easy/flood_fill.js
// https://leetcode.com/problems/flood-fill/

var floodFill = function (image, sr, sc, color) {
  if (image[sr][sc] === color) return image;
  fill(image, sr, sc, color, image[sr][sc]);
  return image;
};

const fill = (image, sr, sc, color, oldColor) => {
  if (sr < 0 || sc < 0 || sr >= image.length || sc > image[0].length || image[sr][sc] !== oldColor) return;

  image[sr][sc] = color;
  fill(image, sr + 1, sc, color, oldColor);
  fill(image, sr - 1, sc, color, oldColor);
  fill(image, sr, sc + 1, color, oldColor);
  fill(image, sr, sc - 1, color, oldColor);
}

File: easy/flood_fill.rb
# https://leetcode.com/problems/flood-fill/

def flood_fill(image, sr, sc, color)
  return image if image[sr][sc] == color
  fill(image, sr, sc, color, image[sr][sc])
  image
end

def fill(image, sr, sc, color, old_color)
  return if (sr >= image.length || sr < 0 || sc >= image[0].length || sc < 0)
  return if image[sr][sc] != old_color

  image[sr][sc] = color
  fill(image, sr + 1, sc, color, old_color)
  fill(image, sr - 1, sc, color, old_color)
  fill(image, sr, sc + 1, color, old_color)
  fill(image, sr, sc - 1, color, old_color)
end
